scripts/knot_benchmark/src/benchmark/cases.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import annotations
+
+import abc
+import enum
+import os
+import shutil
+from pathlib import Path
+import subprocess
+import sys
+import logging
+
+from benchmark import Command
+from benchmark.projects import Project
+
+
+class Benchmark(enum.Enum):
+    """Enumeration of the benchmarks to run."""
+
+    COLD = "cold"
+    """Cold check of an entire project without a cache present."""
+
+    WARM = "warm"
+    """Re-checking the entire project without any changes"."""
+
+    INCREMENTAL = "incremental"
+    """Incremental check between two commits."""
+
+
+def which_tool(name: str) -> Path:
+    tool = shutil.which(name)
+
+    assert (
+        tool is not None
+    ), f"Tool {name} not found. Run the script with `uv run <script>`."
+
+    return Path(tool)
+
+
+class Tool(abc.ABC):
+    def command(
+        self, benchmark: Benchmark, project: Project, venv: Venv
+    ) -> Command | None:
+        """Generate a command to benchmark a given tool."""
+        match benchmark:
+            case Benchmark.COLD:
+                return self.cold_command(project, venv)
+            case Benchmark.WARM:
+                return self.warm_command(project, venv)
+            case Benchmark.INCREMENTAL:
+                return self.incremental_command(project, venv)
+            case _:
+                raise ValueError(f"Invalid benchmark: {benchmark}")
+
+    @abc.abstractmethod
+    def cold_command(self, project: Project, venv: Venv) -> Command: ...
+
+    def warm_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+    def incremental_command(self, project: Project, venv: Venv) -> Command | None:
+        return None
+
+
+class Knot(Tool):
+    path: Path
+    name: str
+
+    def __init__(self, *, path: Path | None = None):
+        self.name = str(path) or "knot"
+        self.path = path or (
+            Path(__file__)
+            .joinpath(
+                "../../../../../",
+                "target",
+                "release",
+                "red_knot",
+            )
+            .resolve()
+        )
```

Comment:
```suggestion
        if path is None:
            path = Path(__file__) / "../../../../../target/release/red_knot"
            self.path = path.resolve()
        else:
            self.path = path
```

Reply:
I'm okay with the `/` stuff but I prefer the inline `or` (I'm a huge fan of single assignments)

Reply:
Fair enough!